### PR TITLE
Remove Invariant_env, empty invariant checks, etc.

### PR DESCRIPTION
**Pull request by mshinwell**
This removes the old invariant-checking infrastructure using `Invariant_env`.  It was never finished and I think we can make do without it.  The main thing we should add somewhere is a check to make sure traps are balanced correctly, but that could be done at Cmm time potentially.

This patch needs some review because I've kept some of the old invariant checks but put them directly into `create` functions and so forth.

### CODE REVIEW

File: middle_end/flambda2/terms/bound_symbols.ml [hunk 1]
```diff
@@ -120,7 +128,27 @@ type t = Pattern.t list
...
+    |> List.concat
+  in
+  let everything_being_defined_as_set =
+    Code_id_or_symbol.Set.of_list everything_being_defined
+  in
+  if List.compare_length_with everything_being_defined
+       (Code_id_or_symbol.Set.cardinal everything_being_defined_as_set)
+     <> 0
+  then
+    Misc.fatal_errorf "Illegal pattern list:@ %a"
```
Comment (Gbury) [inline middle_end/flambda2/terms/bound_symbols.ml:144]: Maybe you could specify here why the pattern list is illegal ? Namely that there is a repeated symbol being bound.

**Review by Gbury — APPROVED** [review]

File: middle_end/flambda2/terms/apply_expr.ml [hunk 2]
```diff
@@ -101,43 +99,22 @@ let [@ocamlformat "disable"] print ppf { callee; continuation; exn_continuation;
...
-    | Function Indirect_unknown_arity ->
-      E.check_simples_are_bound_and_of_kind env args K.value
-    | Function (Indirect_known_arity { param_arity; return_arity = _ }) ->
-      ignore (param_arity : Flambda_arity.With_subkinds.t);
-      E.check_simples_are_bound env args
-    | Method { kind; obj } ->
-      ignore (kind : Call_kind.method_kind);
-      E.check_simple_is_bound_and_of_kind env obj K.value;
-      E.check_simples_are_bound_and_of_kind env args K.value
+    | Function _ | Method _
```
Comment (Gbury) [inline middle_end/flambda2/terms/apply_expr.ml:115]: Why is this case shared with the `C_call` actually ? It seems to be the cause of the CI failure

Comment (mshinwell) [reply middle_end/flambda2/terms/apply_expr.ml]: Fixed, thanks

File: middle_end/flambda2/terms/bound_symbols.ml (hunk 1 above)
Comment (mshinwell) [reply middle_end/flambda2/terms/bound_symbols.ml]: Fixed

**Pull request merged**